File: app/src/main/java/com/wlmtest/forecasttestjava/viewmodel/MainViewModel.java
package com.wlmtest.forecasttestjava.viewmodel;

import android.location.Location;
import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModel;
import com.wlmtest.forecasttestjava.base.events.InternetDisconnectedEvent;
import com.wlmtest.forecasttestjava.model.api.currentweatherdata.CurrentWeatherData;
import com.wlmtest.forecasttestjava.model.api.fivedaysforecastdata.FiveDaysForecastData;
import com.wlmtest.forecasttestjava.model.pojo.CityForecastPojo;
import com.wlmtest.forecasttestjava.model.pojo.forecastfivedays.FiveDaysForecastPojo;
import com.wlmtest.forecasttestjava.repository.ApiDataParser;
import com.wlmtest.forecasttestjava.repository.ForecastTestJavaRepository;
import com.wlmtest.forecasttestjava.view.main.MainOrchestrationHelper;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import javax.inject.Inject;
import java.util.ArrayList;


public class MainViewModel extends ViewModel {

    @Inject
    public MainViewModel(@NonNull ForecastTestJavaRepository forecastTestJavaRepository) {
        super();
        this.forecastTestJavaRepository = forecastTestJavaRepository;
        EventBus.getDefault().register(this);
        setObservers();
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        EventBus.getDefault().unregister(this);
    }

    private ForecastTestJavaRepository forecastTestJavaRepository;


    /**
     * API DATA
     */
    private MutableLiveData<CurrentWeatherData> currentWeatherDataMutableLiveData = new MutableLiveData<>();
    private MutableLiveData<FiveDaysForecastData> fiveDaysForecastDataMutableLiveData = new MutableLiveData<>();

    /**
     * CITY FORECAST
     */
    public MutableLiveData<CityForecastPojo> cityForecastPojoSelectedMutableLiveData = new MutableLiveData<>();
    public MutableLiveData<java.util.List<CityForecastPojo>> cityForecastSearchListMutableLiveData = new MutableLiveData<>();

    /**
     * CITY FORECAST FIVE DAYS
     */
    public MutableLiveData<FiveDaysForecastPojo> fiveDaysForecastPojoMutableLiveData = new MutableLiveData<>();

    /**
     * VISUAL ORCHESTRATION
     */
    public MutableLiveData<String> fragmentOrchestrationHelperMutableLiveData = new MutableLiveData<>();
    public MutableLiveData<Boolean> appInitialized = new MutableLiveData<>();
    public MutableLiveData<Boolean> showInternetUnavailableMessage = new MutableLiveData<>();


    /**
     * Eventbus message handler responsible for receive the emitted event from internet disconnection interceptor.
     *
     * @param event
     */
    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMessageEvent(InternetDisconnectedEvent event) {
        showInternetUnavailableMessage.setValue(true);
    }

    ;


    /**
     * Set all observers for VM
     */
    private void setObservers() {


        /**
         * Responsible for observing the data returned form the API for current weather search.
         */
        this.currentWeatherDataMutableLiveData.observeForever(new Observer<CurrentWeatherData>() {

            @Override
            public void onChanged(CurrentWeatherData currentWeatherData) {

                cityForecastSearchListMutableLiveData.setValue(new ArrayList<CityForecastPojo>());

                if (currentWeatherData != null && currentWeatherData.getList() != null && currentWeatherData.getList().size() >= 0) {

                    java.util.List<CityForecastPojo> cityForecastPojoList = ApiDataParser.parseCurrentWeatherDataToCityForecastPojoList(currentWeatherData);


                    if (cityForecastPojoList.size() == 1) {
                        cityForecastPojoSelectedMutableLiveData.setValue(cityForecastPojoList.get(0));
                        fragmentOrchestrationHelperMutableLiveData.setValue(MainOrchestrationHelper.INSTANCE.getSHOW_FORECAST_FRAGMENT());

                    } else {

                        cityForecastSearchListMutableLiveData.setValue(cityForecastPojoList);
                        fragmentOrchestrationHelperMutableLiveData.setValue(MainOrchestrationHelper.INSTANCE.getSHOW_SEARCH_RESULT_FRAGMENT());

                    }

                } else {
                    fragmentOrchestrationHelperMutableLiveData.setValue(MainOrchestrationHelper.INSTANCE.getSHOW_SEARCH_RESULT_FRAGMENT());
                }

                //
                appInitialized.setValue(true);
            }
        });

        /**
         *
         * Responsible for observing when some Forecast Pojo is selected. This selection will trigger the request of FiveDaysForecastData.
         *
         */
        cityForecastPojoSelectedMutableLiveData.observeForever(new Observer<CityForecastPojo>() {
            @Override
            public void onChanged(CityForecastPojo cityForecastPojo) {

                forecastTestJavaRepository
                        .findFiveDaysForecastData(cityForecastPojo.getCityId(), fiveDaysForecastDataMutableLiveData);

            }
        });

        /**
         * Responsible for observing the API Five days forecast data and trigger the parse to a POJO.
         *
         */
        fiveDaysForecastDataMutableLiveData.observeForever(new Observer<FiveDaysForecastData>() {
            @Override
            public void onChanged(FiveDaysForecastData fiveDaysForecastData) {

                fiveDaysForecastPojoMutableLiveData.setValue(ApiDataParser.parseFiveDaysForecastDataToFiveDaysForecastPojo(fiveDaysForecastData));

            }
        });
    }


    /**
     * Find the weather date from API by name.
     *
     * Responsible for connect the view's request to the repository.
     *
     * @param cityName
     */
    public void findCurrentWeatherDataByCityName(String cityName) {

        forecastTestJavaRepository
                .findCurrentWeatherData(cityName, currentWeatherDataMutableLiveData);

    }

    /**
     * Find the weather date from API by location.
     *
     * Responsible for connect the view's request to the repository.
     *
     * @param location
     */
    public void findCurrentWeatherDataByGPSCoordinates(Location location) {

        forecastTestJavaRepository
                .findCurrentWeatherData(location, currentWeatherDataMutableLiveData);

    }


    /**
     * Set the City Forecast by POJO
     *
     * @param cityForecastPojo
     */
    public void setCityForecastPojoSelected(CityForecastPojo cityForecastPojo) {
        cityForecastPojoSelectedMutableLiveData.setValue(cityForecastPojo);
        fragmentOrchestrationHelperMutableLiveData.setValue(MainOrchestrationHelper.INSTANCE.getSHOW_FORECAST_FRAGMENT());
    }


}

File: app/src/main/java/com/wlmtest/forecasttestjava/viewmodel/MainViewModel.kt
package com.wlmtest.forecasttestjava.viewmodel

import android.location.Location
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.wlmtest.forecasttestjava.base.events.InternetDisconnectedEvent
import com.wlmtest.forecasttestjava.base.events.ProgressDialogLoadingDataEvent
import com.wlmtest.forecasttestjava.model.api.currentweatherdata.CurrentWeatherData
import com.wlmtest.forecasttestjava.model.api.fivedaysforecastdata.FiveDaysForecastData
import com.wlmtest.forecasttestjava.model.pojo.CityForecastPojo
import com.wlmtest.forecasttestjava.model.pojo.forecastfivedays.FiveDaysForecastPojo
import com.wlmtest.forecasttestjava.repository.ApiDataParser
import com.wlmtest.forecasttestjava.repository.ForecastTestKotlinRepository
import com.wlmtest.forecasttestjava.view.main.MainOrchestrationHelper
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

import java.util.ArrayList


class MainViewModel
constructor(private val forecastTestKotlinRepository: ForecastTestKotlinRepository) : ViewModel() {


    /**
     * API DATA
     */
    private val currentWeatherData = MutableLiveData<CurrentWeatherData>()
    private val fiveDaysForecastData = MutableLiveData<FiveDaysForecastData>()

    /**
     * CITY FORECAST
     */
    var cityForecastPojoSelected = MutableLiveData<CityForecastPojo>()
    var cityForecastSearchList = MutableLiveData<List<CityForecastPojo>>()

    /**
     * CITY FORECAST FIVE DAYS
     */
    var fiveDaysForecastPojo = MutableLiveData<FiveDaysForecastPojo>()

    /**
     * VISUAL ORCHESTRATION
     */
    var fragmentOrchestrationHelper = MutableLiveData<String>()
    var appInitialized = MutableLiveData<Boolean>()
    var showInternetUnavailableMessage = MutableLiveData<Boolean>()

    var progressDialogLoadingData= MutableLiveData<Boolean>();

    init {
        EventBus.getDefault().register(this)
        setObservers()
    }

    override fun onCleared() {
        super.onCleared()
        EventBus.getDefault().unregister(this)
    }


    /**
     * Eventbus message handler responsible for receive the emitted event from internet disconnection interceptor.
     *
     * @param event
     */
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(event: InternetDisconnectedEvent) {
        showInternetUnavailableMessage.value = true
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(event: ProgressDialogLoadingDataEvent) {
        progressDialogLoadingData.value = event.loading
    }

    /**
     * Set all observers for VM
     */
    private fun setObservers() {


        /**
         * Responsible for observing the data returned form the API for current weather search.
         */
        currentWeatherData.observeForever { currentWeatherData ->
            cityForecastSearchList.value = ArrayList()

            if (currentWeatherData != null && currentWeatherData.list != null && currentWeatherData.list!!.size >= 0) {

                val cityForecastPojoList =
                    ApiDataParser.parseCurrentWeatherDataToCityForecastPojoList(currentWeatherData)


                if (cityForecastPojoList.size == 1) {
                    cityForecastPojoSelected.value = cityForecastPojoList[0]
                    fragmentOrchestrationHelper.setValue(MainOrchestrationHelper.SHOW_FORECAST_FRAGMENT)

                } else {

                    cityForecastSearchList.value = cityForecastPojoList
                    fragmentOrchestrationHelper.setValue(MainOrchestrationHelper.SHOW_SEARCH_RESULT_FRAGMENT)

                }

            } else {
                fragmentOrchestrationHelper.setValue(MainOrchestrationHelper.SHOW_SEARCH_RESULT_FRAGMENT)
            }

            //
            appInitialized.value = true
        }

        /**
         *
         * Responsible for observing when some Forecast Pojo is selected. This selection will trigger the request of FiveDaysForecastData.
         *
         */
        cityForecastPojoSelected.observeForever { cityForecastPojo ->
            forecastTestKotlinRepository
                .findFiveDaysForecastData(cityForecastPojo.cityId!!, fiveDaysForecastData)
        }

        /**
         * Responsible for observing the API Five days forecast data and trigger the parse to a POJO.
         *
         */
        fiveDaysForecastData.observeForever { fiveDaysForecastData ->
            fiveDaysForecastPojo.value =
                ApiDataParser.parseFiveDaysForecastDataToFiveDaysForecastPojo(fiveDaysForecastData)
        }
    }


    /**
     * Find the weather date from API by name.
     *
     * Responsible for connect the view's request to the repository.
     *
     * @param cityName
     */
    fun findCurrentWeatherDataByCityName(cityName: String) {

        forecastTestKotlinRepository
            .findCurrentWeatherData(cityName, currentWeatherData)

    }

    /**
     * Find the weather date from API by location.
     *
     * Responsible for connect the view's request to the repository.
     *
     * @param location
     */
    fun findCurrentWeatherDataByGPSCoordinates(location: Location) {

        forecastTestKotlinRepository
            .findCurrentWeatherData(location, currentWeatherData)

    }


    /**
     * Set the City Forecast by POJO
     *
     * @param cityForecastPojo
     */
    fun setCityForecastPojoSelected(cityForecastPojo: CityForecastPojo) {
        //clearn  5days data
        fiveDaysForecastPojo.value = null

        cityForecastPojoSelected.value = cityForecastPojo
        fragmentOrchestrationHelper.value = MainOrchestrationHelper.SHOW_FORECAST_FRAGMENT
    }


}
